File: src/test/java/xyz/realmath/AbelianGroupTest.java
package xyz.realmath;

import static org.junit.Assert.assertSame;

import org.junit.Test;

public class AbelianGroupTest {
  private static final TestClass T1 = new TestClass();
  private static final TestClass T2 = new TestClass();
  private static final TestClass T3 = new TestClass();
  private static final TestClass T4 = new TestClass();

  private static final class TestClass implements AbelianGroup<TestClass> {

    @Override
    public TestClass add(TestClass t) {
      if (this == T1 && t == T3) {
        return T4;
      }
      return null;
    }

    @Override
    public TestClass negate() {
      if (this == T2) {
        return T3;
      }
      return null;
    }
  }

  @Test
  public void subtract() {
    assertSame(T4, T1.subtract(T2));
  }
}

File: src/test/kotlin/xyz/realmath/AbelianGroupTest.kt
package xyz.realmath

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kotlin.test.Test
import kotlin.test.assertTrue

class AbelianGroupTest {
    @Test
    fun <T : AbelianGroup<T>> isZero() {
        @Suppress("UNCHECKED_CAST")
        val ringElm = mockk<AbelianGroup<T>>() as T
        every { ringElm.zero() } returns ringElm

        assertTrue(ringElm.isZero)
    }

    @Test
    fun <T : AbelianGroup<T>> minus() {
        @Suppress("UNCHECKED_CAST")
        val ringElm = mockk<AbelianGroup<T>>(relaxed = true) as T
        @Suppress("UNCHECKED_CAST")
        val negRingElm = mockk<AbelianGroup<T>>(relaxed = true) as T
        every { ringElm.unaryMinus() } returns negRingElm

        ringElm - ringElm

        verify { ringElm.unaryMinus() }
        verify { ringElm.plus(negRingElm) }
    }
}
